app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.tsx

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import React from "react";
+import { RouteComponentProps, useParams, Link } from "@reach/router";
+import Container from "react-bootstrap/Container";
+import Row from "react-bootstrap/Row";
+import Col from "react-bootstrap/Col";
+import Nav from "react-bootstrap/Nav";
+import { BASE_PATH } from "../../lib/constants";
+
+type AppLayoutWithSidebarProps = {
+  testid?: string;
+  children: React.ReactNode;
+} & RouteComponentProps;
+
+export const AppLayoutWithSidebar = ({
+  children,
+  testid = "AppLayoutWithSidebar",
+}: AppLayoutWithSidebarProps) => {
+  // an empty string makes Storybook happy
+  const { slug } = useParams() || "";
```

Comment:
The only downside to this is we're considered to not having tested a branch here but it should only ever hit the _or_  in Storybook. @lmorchard I was scratching my head trying to get Storybook to be happy without this, [storybook-react-router](https://www.npmjs.com/package/storybook-react-router) won't work for us since we're using reach-router.

I thought we were set to 100% coverage on _everything_ - looks like an uncovered line squicked through CI on `useExitWarning` so I'mma see what happens here?

![image](https://user-images.githubusercontent.com/13018240/96938497-2b588600-1490-11eb-8acf-0bcc175e2194.png)

Reply:
So, I did some exploratory tinkering here and came up with this bit of mocking for the story that constructs just enough of a router to make `useParams` happy. 

Maybe this will help in other cases that depend on mocking router hooks?

```diff
diff --git a/app/experimenter/nimbus-ui/src/components/App/index.tsx b/app/experimenter/nimbus-ui/src/components/App/index.tsx
index 2159f5b8..30800fdb 100644
--- a/app/experimenter/nimbus-ui/src/components/App/index.tsx
+++ b/app/experimenter/nimbus-ui/src/components/App/index.tsx
@@ -5,7 +5,6 @@
 import React from "react";
 import { Router, Redirect, RouteComponentProps } from "@reach/router";
 
-import { BASE_PATH } from "../../lib/constants";
 import PageHome from "../PageHome";
 import PageNew from "../PageNew";
 import PageEditOverview from "../PageEditOverview";
diff --git a/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.stories.tsx b/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.stories.tsx
index 238fd207..b935279c 100644
--- a/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.stories.tsx
+++ b/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.stories.tsx
@@ -4,17 +4,42 @@
 
 import React from "react";
 import { storiesOf } from "@storybook/react";
-import { LocationProvider } from "@reach/router";
+import {
+  createMemorySource,
+  createHistory,
+  LocationProvider,
+  Router,
+  RouteComponentProps,
+} from "@reach/router";
 import { withLinks } from "@storybook/addon-links";
 import AppLayoutWithSidebar from ".";
 
 storiesOf("components/AppLayoutWithSidebar", module)
   .addDecorator((getStory) => <LocationProvider>{getStory()}</LocationProvider>)
   .addDecorator(withLinks)
-  .add("basic", () => (
-    <div data-testid="app">
-      <AppLayoutWithSidebar>
-        <p>App contents go here</p>
-      </AppLayoutWithSidebar>
-    </div>
-  ));
+  .add("basic", () => <Subject />);
+
+const Subject = ({ path = "/demoslug/edit" }) => {
+  let source = createMemorySource(path);
+  let history = createHistory(source);
+
+  return (
+    <LocationProvider history={history}>
+      <Router>
+        <Route
+          path=":slug/edit"
+          data-testid="app"
+          component={() => (
+            <AppLayoutWithSidebar>
+              <p>App contents go here</p>
+            </AppLayoutWithSidebar>
+          )}
+        />
+      </Router>
+    </LocationProvider>
+  );
+};
+
+const Route = (
+  props: { component: () => React.ReactNode } & RouteComponentProps,
+) => <div {...{ props }}>{props.component()}</div>;
diff --git a/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.tsx b/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.tsx
index 1d516e3d..2d0ac82c 100644
--- a/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.tsx
+++ b/app/experimenter/nimbus-ui/src/components/AppLayoutWithSidebar/index.tsx
@@ -20,7 +20,7 @@ export const AppLayoutWithSidebar = ({
   testid = "AppLayoutWithSidebar",
 }: AppLayoutWithSidebarProps) => {
   // an empty string makes Storybook happy
-  const { slug } = useParams() || "";
+  const { slug } = useParams();
   return (
     <Container fluid className="h-100vh" data-testid={testid}>
       <Row className="h-md-100">
```
```

Reply:
Yeee thanks Les, that's great. I took this and made a `RouterSlugProvider` and added it to `test-utils.tsx`... I originally stuck it in `storybook-utils.tsx` but I suppose it could maybe be used elsewhere? Can move if preferred.

I'm going to file a ticket for this tomorrow - I didn't realize CI wasn't set up to fail based on line coverage. We're going to need some in `mocks.tsx` and these new lines in `test-utils.tsx`.

Reply:
I'd suggest maybe we exclude mocks & test-utils from test coverage metrics. IMO, only production code should have coverage requirements enforced